### PR TITLE
Add support for internal module to Companion

### PR DESCRIPTION
**Pull request by elecpower**
Based on ~~main~~ Companion YAML branch ~~as not sure where yaml would be at and it compiles~~.

Things to do:
- [x] finish boards module valid combinations bool ModuleData::isInternalModuleAvailable(int moduleType)
- [ ] on change of internal module type update each model if protocol not supported by module
- [x] yaml for field post merge
- [x] set default internal module where appropriate (conversion from older formats, new file, etc)

### CODE REVIEW

File: companion/src/firmwares/boards.cpp [hunk 1]
```diff
@@ -486,6 +486,9 @@ int Boards::getCapability(Board::Type board, Board::Capability capability)
     case NumFunctionSwitches:
       return IS_JUMPER_TPRO(board) ? 6 : 0;
 
+    case HasInternalModuleSupport:
+      return (IS_STM32(board) && !IS_TARANIS_SMALL(board));
```
Comment (raphaelcoeffic) [inline companion/src/firmwares/boards.cpp:490]: Should be `!IS_TARANIS_X9()`. `IS_TARANIS_SMALL()` has all the 128x64 which do have support for configurable internal module.

File: companion/src/firmwares/moduledata.cpp [hunk 2]
```diff
@@ -262,3 +268,274 @@ int ModuleData::getMaxChannelCount()
...
+//  static
+QString ModuleData::typeToString(int type)
+{
+  static const char * strings[] = {
+    "OFF",
+    "PPM",
+    "FrSky XJT",
+    "ISRM",
+    "DSM2",
+    "TBS Crossfire",
```
Comment (raphaelcoeffic) [inline companion/src/firmwares/moduledata.cpp:344]: CRSF is not just TBS. ELRS uses CRSF as well, which could be confusing for users.

File: companion/src/firmwares/moduledata.cpp [hunk 3]
```diff
@@ -262,3 +268,274 @@ int ModuleData::getMaxChannelCount()
...
+bool ModuleData::isInternalModuleAvailable(int moduleType)
+{
+  if (moduleType == MODULE_TYPE_NONE)
+    return true;
+
+  Firmware *fw = getCurrentFirmware();
+  Board::Type board = fw->getBoard();
+  QString id = fw->getId();
+
+  if (!Boards::getCapability(board, Board::HasInternalModuleSupport))
```
Comment (raphaelcoeffic) [inline companion/src/firmwares/moduledata.cpp:370]: Whereby the setting is not "configurable", these radios still have an internal module (which is fixed).

**Comment by raphaelcoeffic** [comment]
> on change of internal module type update each model if protocol not supported by module

@elecpower Whatever you choose, I'd prefer not having the models being changed silently, as this potentially modifies every module setting in every model. If possible, some kind of warning/choice would be nice.

**Comment by elecpower** [comment]
> > on change of internal module type update each model if protocol not supported by module
> 
> @elecpower Whatever you choose, I'd prefer not having the models being changed silently, as this potentially modifies every module setting in every model. If possible, some kind of warning/choice would be nice.

Okay a warning it is at this stage. My initial attempts to update each model were not successful due to the way they are loaded by the midi dialog and the effect of Qt events. Will revisit with choice to update later.

**Comment by pfeerick** [comment]
That sounds good... and for the moment is the same the radio, as the radio doesn't change the setting either. 

Just tried this with two TX16Ss with different internal module configurations, and worked perfectly for both. One minor more... cosmetic issue though... when changing from say say MULTI to XJT ... first the model page looks like this with MULTI, and it looks like this after having made the change to XJT, and re-opening it. Perhaps it should be OFF? As currently if you do write that config it will write invalid settings - i.e `TYPE_MULTIMODULE` - so is just at odds with what is displayed. 

When set to MULTI:
image

When set to something else:
image

**Comment by elecpower** [comment]
When changed from MULTI the protocol MULTI is not in the drop down list thus the blank. However the ui logic references the hidden MULTI protocol (which as you point out is also written) and builds the display accordingly.

I'm not in favour of keeping invalid protocols after changing module. IMHO it is far better to set to valid value of OFF as there could be issues if the radio doesn't check for invalid protocols, and as you note it is confusing to users. 

If not setting to OFF on change of internal module and the user does not visit each model to get prompted, then saves the settings they will have invalid data saved.

So I vote for a warning message and (if I can get it working) advising all invalid protocols will be set to OFF and in the interim when editing the mode set to OFF if invalid.

**Comment by elecpower** [comment]
@pfeerick after further discussion with @raphaelcoeffic, yaml is priority and perform a staged improvement by:

- [ ] modifying the warning message on change of internal module to allow the user the option to undo the change;
- [ ] updating invalid protocols to OFF on model edit load (fixes ui and saved values) and warning message to user;
- [ ] updating invalid protocols for all models to OFF at time of changing the internal module and update the warning message.

Thoughts?

**Comment by pfeerick** [comment]
Sounds good. I would like the see the last one more as a question - i.e. you have the option of updating them all, but it's not required, especially as no. 2 would catch that state, but that is just my view at present ;) 

Another point to consider is the default radio level internal module setting... it currently defaults to blank on a new profile... intentional or accidental? If the latter, something else to go on the todo list, as in the current form I think this should go in, and we can hopefully get yaml companion merged in, and carry on with incremental improvements in another round.

**Comment by elecpower** [comment]
> Another point to consider is the default radio level internal module setting... it currently defaults to blank on a new profile... intentional or accidental? If the latter, something else to go on the todo list, as in the current form I think this should go in [edit: unless you want to implement no. 1 as part of this PR?], and we can hopefully get yaml companion merged in, and carry on with incremental improvements in another round.

Latest commits set default

**Comment by elecpower** [comment]
> Sounds good. I would like the see the last one more as a question - i.e. you have the option of updating them all, but it's not required, especially as no. 2 would catch that state, but that is just my view at present ;)

We can go with a question and see what feedback we get

**Pull request merged**